### ISSUE TITLE
reattach-to-user-namespace doesn't seem to work for tmux commands when tmux is started directly (and not from a shell)

### ISSUE THREAD

**Issue by jeroenj** [ChrisJohnsen/tmux-MacOSX-pasteboard#58]
I have Terminal.app set up to start tmux directly instead of a default shell:
![screenshot 2017-01-10 20 16 42](https://cloud.githubusercontent.com/assets/76424/21820974/0b446f78-d772-11e6-8076-e3c81fe3de85.png)

This means that if tmux is quit the Terminal window will be closed too. This also makes it so that all of my Terminal windows open the same tmux session (by default).

I have the following in my `.tmux.conf`:

```
set -g default-shell $SHELL
set -g default-command "reattach-to-user-namespace -l ${SHELL}"

unbind -t vi-copy Enter
bind-key -t vi-copy Enter copy-pipe "pbcopy"
bind-key -t vi-copy y copy-pipe "pbcopy"
```

I have reattach-to-user-namespace (2.5) installed with the `--with-wrap-pbcopy-and-pbpaste` option. I've also got the latest tmux (2.3) installed.

It works perfectly when calling the `pbcopy` wrapper directly from my shell (inside tmux), copying in vim, etc. However when using copy mode in tmux itself it seems that reattach-to-user-namespace doesn't do anything. It works silently. I've debugged this by adding a custom wrapper script to debug that `copy-pipe` is working correctly as the buffer gets passed to the command as expected.

When I check `Run inside shell` in the Terminal Shell preferences (above screenshot) copying from tmux copy mode is working as expected too. This makes me think that there is some issue with it if tmux isn't started from a shell itself.

I skimmed through the source code but C isn't my strong point and I couldn't think of any reason at first sight. For now using `Run inside shell` makes it work but I'd like to disable it again as it means running one less shell. Enabling it also breaks my flow with multiple Terminal windows as when it is enabled the tmux sessions remain open when closing the Terminal windows.

**Comment by ChrisJohnsen** [comment]
Where does Homebrew(?) install those wrapped *pbcopy* and *pbpaste*? Are you sure your `copy-pipe` command is using the wrapped versions?

Using bare command names in `copy-pipe` will end up using the *tmux* server’s PATH (`tmux show-environment -g PATH`), which may not include the wrapped versions but likely does include the system (non-wrapped) versions.

You might try setting `copy-pipe` to use the full path to the wrapped version of *pbcopy*.

GUI apps usually have a minimal PATH (e.g. `/usr/bin:/bin:/usr/sbin:/sbin`), so a *tmux* server started directly by *Terminal* would inherit that basic PATH and not “see” the wrapped version of *pbcopy*. On the other hand, a *tmux* server started from one of your “usual” shells will inherit a PATH that has probably been extended (by your shell’s startup configuration) to include the wrapped *pbcopy* (i.e. the directory where your wrapped version is installed).

**Comment by jeroenj** [comment]
They're installed as `/usr/local/bin/pbcopy` and `/usr/local/bin/pbaste`. I forgot to mention it in the issue but I've also tried using them directly. I've also tried using `reattach-to-user-namespace /usr/bin/pbocpy` but none of them worked.

I ended up adding a custom script with the following code when debugging:

```
#!/bin/sh

cat - > /tmp/buffer

echo foo | /usr/local/bin/pbcopy
reattach-to-user-namespace "echo bar | /usr/local/bin/pbcopy"
reattach-to-user-namespace "echo baz | /usr/bin/pbcopy"
```

Nor `foo`, `bar` or `baz` was added to the clipboard. When calling that script directly from a (zsh) shell within tmux works as expected. The contents of tmux's buffer was written in `/tmp/buffer` as expected.

**Comment by ChrisJohnsen** [comment]
You might try putting `exec >/tmp/copy-pipe.out 2>&1` at the top of that debugging script; having that in there will let you see any stdout or stderr output from the shell running that script and the executables it runs.

I would have expected the `foo` test to work, but the `bar` and `baz` tests are faulty. The wrapper uses *exec*, not *system*, so giving it shell-based command strings will not do what you want. Try (e.g.)

* `reattach-to-user-namespace sh -c "echo baz | /usr/bin/pbcopy"`, 
* or (without and extra shell invocation) `echo baz | reattach-to-user-namespace /usr/bin/pbcopy` instead.

You might also need to use an absolute path to the wrapper since the PATH may not include `/usr/local/bin` (if you capture to `copy-pipe.out`, you can add `echo "PATH: $PATH"` to see its value in the script; also you would probably see something like “command not found” from the shell).

**Comment by jeroenj** [comment]
Thanks. I didn't know about `exec >/tmp/copy-pipe.out 2>&1`. By adding that I found out that it indeed couldn't find `reattach-to-user-namespace` in the provided `PATH` as it only contained `/usr/bin:/bin`.

Instead of using the `pbcopy` wrapper provided by Homebrew (which uses `reattach-to-user-namespace` without providing an absolute `PATH` I've switched to using both tools directly with their absolute paths:

```
bind-key -t vi-copy Enter copy-pipe "/usr/local/bin/reattach-to-user-namespace /usr/bin/pbcopy"
bind-key -t vi-copy y copy-pipe "/usr/local/bin/reattach-to-user-namespace /usr/bin/pbcopy"
```

Thanks a lot for helping me investigate this. It helped me to better understand some of the things going on. :) 👍
